Sources/NIOCore/AsyncSequences/NIOBackPressuredAsyncSequence.swift

```diff
@@ -0,0 +1,1004 @@
+            switch self.state {
+            case .initial:
+                // Nothing was yielded nor did anybody call next
+                // This means we can just transition to finished
+                self.state = .finished
+
+                return .callDidTerminate
+
+            case .streaming(_, let buffer, .some(let continuation), _, _):
+                // We have a continuation, this means our buffer must be empty
+                // Furthermore, we can now transition to finished
+                // and resume the continuation with `nil`
+                precondition(buffer.isEmpty)
+
+                self.state = .finished
+
+                return .resumeContinuationWithNilAndCallDidTerminate(continuation)
+
+            case .streaming(_, let buffer, continuation: .none, _, let iteratorInitialized):
+                if buffer.isEmpty {
+                    // There is nothing left in the buffer and we can transition to finished
+                    self.state = .finished
+
+                    return .callDidTerminate
+                } else {
+                    // There is some stuff left in the buffer and we have to continue to unbuffer
+                    self.state = .sourceFinished(
+                        buffer: buffer,
+                        iteratorInitialized: iteratorInitialized
+                    )
+
+                    return .none
+                }
+
+            case .sourceFinished, .finished:
+                // If the source has finished, finishing again has no effect.
+                return .none
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `cancelled()`.
+        @usableFromInline
+        enum CancelledAction {
+            /// Indicates that ``NIOBackPressuredAsyncSequenceDelegate/didTerminate()`` should be called.
+            case callDidTerminate
+            /// Indicates that the continuation should be resumed with `nil` and
+            /// that ``NIOBackPressuredAsyncSequenceDelegate/didTerminate()`` should be called.
+            case resumeContinuationWithNilAndCallDidTerminate(CheckedContinuation<Element?, Never>)
+            /// Indicates that nothing should be done.
+            case none
+        }
+
+        @inlinable
+        mutating func cancelled() -> CancelledAction {
+            switch self.state {
+            case .initial:
+                // This can happen if the `Task` that calls `next()` is already cancelled.
+                self.state = .finished
+
+                return .callDidTerminate
+
+            case .streaming(_, _, .some(let continuation), _, _):
+                // We have an outstanding continuation that needs to resumed
+                // and we can transition to finished here and inform the delegate
+                self.state = .finished
+
+                return .resumeContinuationWithNilAndCallDidTerminate(continuation)
```

Comment:
We can't throw a `CancellationError` here since our `NIOBackPressuredAsyncSequence` is non throwing. AFAIK for the sequences that are not throwing we should resume with `nil` for the throwing counter pat we can throw the `CancellationError`

Reply:
If you can't throw returning nil sounds ok for this... Arguably this might want to get a throwing pair type but you'll be forced to replicate a lot of the type I guess :-(

Reply:
It a non throwing `AsyncSequence` is running in a cancelled `Task` it will return nil at the earliest moment it can. (That may be deferred to the next call of `next()` in some cases since cancellation is cooperative, but a best effort should be made)